docs/fsharp/style-guide/conventions.md

```diff
@@ -111,10 +113,148 @@ This enables the following:
+
+This third point is quite critical. For sufficiently complex operations, failing to use exceptions can result in dealing with structures like this:
+
+```fsharp
+Result<Result<MyType, string>, string list>
+```
+
+Which can easily lead to absurd code like pattern matching on "stringly-typed" errors:
+
+```fsharp
+let result = doStuff()
+match result with
+| Ok r -> ...
+| Error e ->
+    if e.Contains "String1" then ...
+    elif e.Contains "Error string 2" then ...
+    else ... // Who knows?
+```
+
+Additionally, it can be tempting to simply swallow any exception in the desire for a "simple" function which returns a nicer type:
+
+```fsharp
+// This is bad!
+let tryReadAllText (path : string) : string option =
+    try System.IO.File.ReadAllText path |> Some
+    with _ -> None
+```
+
+Unfortunately, `ReadAllText` can throw any number of exceptions based on the strange things which can happen on a file system, and this code throws away any information about what might actually be going wrong in your environment. If you were to replace this code with a result type, then you're back to stringly-typed error message parsing:
+
+```fsharp
+// This is bad!
+let tryReadAllText (path : string) : string option =
+    try System.IO.File.ReadAllText path |> Ok
+    with e -> Error e.Message
+
+let r = tryReadAllText "path-to-file"
+match r with
+| Ok text -> ...
+| Error e ->
+    if e.Contains "uh oh, here we go again..." then ...
+    else ...
+```
+
+And placing the exception in the `Error` constructor just forces you to properly deal with the exception type at the call site rather than in the function.
+
+That said, types such as `Result<'Success, 'Error>` are perfectly fine for basic operations where they aren't nested, and F# optional types are perfect for representing when something could either reutrn *something* or *nothing*. They are not a replacement for exceptions, though, and should not be used in an attempt to replace exceptions.
+
+## Partial application and point-free programming
+
+F# supports partial application, and thus, various ways to program in a point-free style. This can be beneficial for code re-use within a module or the implmentation of something, but it is generally not something to expose publically. In general, point-free programming is not a virtue in and of itself, and comes with a cognitive cost for people who are not familiar with the style.
+
+## Do not use partial application and currying in public APIs
+
+With little exception, the use of partial application in public APIs can be confusing for consumers. Generally speaking, `let`-bound values in F# code are **values**, not **function values**. Mixing together values and function values can result in saving a very small number of lines of code in exchange for quite a bit of cognitive overhead, especially if combined with operators such as `>>` to compose functions together.
```

Comment:
A point to drive through here is that curried functions do not label their arguments, so instead of
```fsharp
val foo : name:string -> age:int -> unit
```
you get
```fsharp
val foo : (string -> int -> unit)
```
which can be confusing from a documentation perspective.

Reply:
Recommended workaround would be to do full-blown eta-expansion, even if it does feel redudant.

Reply:
Covered.